README.md
# Wampum Popups
A lightweight but flexible WordPress popups plugin utilizing [oiubounce](https://github.com/carlsednaoui/ouibounce).
* Use a simple shortcode (or PHP function) to create 1 or more popups (or slideups) throughout your website
* Popup types include exit itent, timed, link, and button
* Automatically works with WordPress galleries (when set to link to file or attachment page)
* Default exit intent and timed popups store an expiring cookie in the browser so it won't continue to display after a user closes it
* Various options allow fine-tuning (see below)
* Easy plugin updates in the WordPress Dashboard via [GitHub Updater plugin](https://github.com/afragen/github-updater)

![Wampum Popups modal example](assets/wampum-popups-modal.jpg)

![Wampum Popups slideup example](assets/wampum-popups-slideup.jpg)

## Basic Usage
Tip: A browser extention like [Cookie Inspector](https://chrome.google.com/webstore/detail/cookie-inspector/jgbbilmfbammlbbhmmgaagdkbkepnijn) is helpful as it lets you manually clear individual cookies 1 at a time

### Shortcode

```
[wampum_popup type="exit" style="modal"] // HTML content here [/wampum_popup]
```

#### Manually launch a popup with a link/button in the WP editor

```
[wampum_popup type="link" text="Click Here"] // HTML content here [/wampum_popup]
```
```
[wampum_popup type="button" text="Click Here"] // HTML content here [/wampum_popup]
```

### PHP functions

Use the `wampum_popups` hook to safely output a popup. This way, if the plugin gets deactivated your popup won't throw errors and/or break your site

```
add_action( 'wampum_popups', 'prefix_do_wampum_popup' );
function prefix_do_wampum_popup() {

	// Bail if not a single post
	if ( ! is_singular('post') ) {
		return;
	}

	$content = '// Some HTML';

	$args = array(
		'type'	=> 'timed',
		'style'	=> 'slideup',
	);
	wampum_popup( $content, $args );

}
```
Optionally use `get_wampum_popup( $content, $args );` to return the popup

**Note:** Scripts will already be enqueued

#### Manually launch a popup with a link/button in your PHP template

```
$content = '// Some HTML';
$args	 = array(
	'type'	=> 'link', (or 'button')
	'text'	=> 'Click Here',
);
wampum_popup_link( $content, $args );
```

Optionally use `get_wampum_popup_link( $content, $args );` to return the link

**Note:** Scripts will already be enqueued and popup will be loaded in the DOM

### WordPress Galleries

Create a gallery and "Link To" either "Attachment Page" (not sure why anyone would ever do this) or "Media File".

Wampum Popups filters the attachement image or permalink URL and uses the 'large' image size in place.

## Shortcode parameters & PHP args

### type (**required**)

(string) 'exit|timed|link|button'

**Default** `null`

Type of popup. This is the only required parameter.

**Note:** 'exit' and 'timed' should be used with `wampum_popup()` hooked into `wampum_popups` action hook, while 'link' and 'button' should be used with `wampum_popup_link()` in your template, functions.php, etc..

---

### close_button

(boolean) true|false

**Default** `true`

Show the close button

---

### close_outside

(boolean) true|false

**Default** `true`

Close popup by clicking outside the modal

---

### style

(string) 'modal|slideup'

**Default** `'modal'`

Style of the popup

---

### text

(string) 'Click Here'

**Default** `null`

The link/button text when using `wampum_popup_link()` function

---

### time

(integer) 4000

**Default** `4000`

Time in milliseconds. 4000ms = 4s

---

### logged_in and logged_out

**Note:** if logged_in and logged_out are both true, the popup will never load... naturally

#### logged_in

(boolean) true|false

**Default** `false`

Show only to logged in users

#### logged_out

(boolean) true|false

**Default** `false`

Show only to logged out users

---

### width

(string) '400px|auto|calc(100vw - 5%)''

**Default** `'400px'`

Max width of the popup, formatted for CSS `max-width: 400px;` spec. Helps for custom designs, larger images, videos, or to make slideups not so weird if too big.

---

## ouibounce args

These args are specific to the [oiubounce](https://github.com/carlsednaoui/ouibounce) script

Most of these won't be used very often, so only a few are documented here

### aggressive

(boolean) true|false

**Default** `false`

Force the popup to show, always. Don't be annoying.

---

### Other available options/parameters

Read about these options on the [ouibounce wiki](https://github.com/carlsednaoui/ouibounce/blob/master/README.md)

* 'callback' (This has been removed/reserved for future use)
* 'cookieExpire'
* 'cookieDomain'
* 'cookieName'
* 'delay'
* 'sensitivity'
* 'sitewide'
* 'timer'

## Advanced Usage

This example showing the ability to use multiple/different popups on the same site.


```
/**
 * Display different popups on different sections of the site
 * 'cookieName' param is used to make the popups act independently
 * if cookieName was the same (this is default behavior) and the user viewed one popup,
 * they wouldn't see the other until the cookie expired.
 *
 * @uses Wampum - Popups plugin
 *
 * @return void
 */
add_action( 'wampum_popups', 'prefix_do_wampum_popups' );
function prefix_do_wampum_popups() {

	// Show this popup on posts
	if ( is_singular('post') ) {
		$content = '<p>My post popup content here</p>';
		$args	 = array(
			'cookieName' => 'prefix_posts_popup_viewed'
			'type'		 => 'exit',
			'style'		 => 'modal',
		);
		wampum_popup( $content, $args );
	}

	// Show this popup on pages
	if ( is_singular('page') ) {
		$content = '<p>My post page content here</p>';
		$args	 = array(
			'cookieName' => 'prefix_pages_popup_viewed'
			'type'		 => 'timed',
			'style'		 => 'slideup',
		);
		wampum_popup( $content, $args );
	}

}
```


CHANGES.md
#### 2.1.0
* Added: Included updater script.

#### 2.0.1
* Update plugin header to point to new repo in bizbudding GitHub organization

#### 2.0.0
* Convert name to match wampum plugin suite
* Full rebuild, with all new shortcodes, a some new parameters

#### 1.0.2
* CSS tweaks to better handle popups with a lot of content

#### 1.0.1
* Fix click in popup causing it to close

#### 1.0.0
* Let's get this party started


slim-popups.php
<?php
/**
 * @package   Slim_Popups_Setup
 * @author    BizBudding, INC <mike@bizbudding.com>
 * @license   GPL-2.0+
 * @link      http://bizbudding.com.com
 * @copyright 2016 BizBudding, INC
 *
 * @wordpress-plugin
 * Plugin Name:        Slim Popups
 * Description: 	   A lightweight developer-based popups plugin utilizing oiubounce
 * Plugin URI:         https://github.com/JiveDig/slim-popups
 * Author:             Mike Hemberger
 * Author URI:         http://bizbudding.com
 * Text Domain:        slim-popups
 * License:            GPL-2.0+
 * License URI:        http://www.gnu.org/licenses/gpl-2.0.txt
 * Version:            1.0.2
 * GitHub Plugin URI:  https://github.com/JiveDig/slim-popups
 * GitHub Branch:	   master
 */


// Exit if accessed directly.
if ( ! defined( 'ABSPATH' ) ) exit;

/**
 * Main function to create a popup
 *
 * Basic usage - Popup content in child theme: /child-theme-name/slim-popups/main-popup.php
 * slim_popup( 'main-popup' );
 *
 * Showing default settings
 * $options = array(
 * 		'css'  			=> true, 	// whether or not to load the stylesheet
 *		'style'			=> 'modal', // 'modal' or 'slideup'
 *		'time'			=> '4000',  // time in milliseconds
 *		'type' 			=> 'exit',  // 'exit' or 'timed'
 * );
 * slim_popup( 'main-popup', $options );
 *
 * @since  1.0.0
 *
 * @param  string  $filename  File name (one word, no hyphens or underscores)
 * @param  array   $options	  Plugin options (css, style, time, type)
 * @param  array   $args   	  Ouibounce object properties. See (https://github.com/carlsednaoui/ouibounce#options)
 *
 * @return void
 */
function slim_popup( $filename, $options = array(), $args = array() ) {
	Slim_Popups()->slim_popup( $filename, $options, $args );
}

if ( ! class_exists( 'Slim_Popups_Setup' ) ) :
/**
 * Main Slim_Popups_Setup Class.
 *
 * @since 1.0.0
 */
final class Slim_Popups_Setup {
	/**
	 * Singleton
	 * @var   Slim_Popups_Setup The one true Slim_Popups_Setup
	 * @since 1.0.0
	 */
	private static $instance;

	public $templates;
	/**
	 * Main Slim_Popups_Setup Instance.
	 *
	 * Insures that only one instance of Slim_Popups_Setup exists in memory at any one
	 * time. Also prevents needing to define globals all over the place.
	 *
	 * @since   1.0.0
	 * @static  var array $instance
	 * @return  object | Slim_Popups_Setup The one true Slim_Popups_Setup
	 */
	public static function instance() {
		if ( ! isset( self::$instance ) ) {
			// Setup the setup
			self::$instance = new Slim_Popups_Setup;
			// Methods
			self::$instance->setup_constants();
			self::$instance->includes();
			self::$instance->setup();
			// Instantiate Classes
			self::$instance->templates = Slim_Popups_Template_Loader::instance();
		}
		return self::$instance;
	}
	/**
	 * Throw error on object clone.
	 *
	 * The whole idea of the singleton design pattern is that there is a single
	 * object therefore, we don't want the object to be cloned.
	 *
	 * @since   1.0.0
	 * @access  protected
	 * @return  void
	 */
	public function __clone() {
		// Cloning instances of the class is forbidden.
		_doing_it_wrong( __FUNCTION__, __( 'Cheatin&#8217; huh?', 'wampum' ), '1.0' );
	}
	/**
	 * Disable unserializing of the class.
	 *
	 * @since   1.0.0
	 * @access  protected
	 * @return  void
	 */
	public function __wakeup() {
		// Unserializing instances of the class is forbidden.
		_doing_it_wrong( __FUNCTION__, __( 'Cheatin&#8217; huh?', 'wampum' ), '1.0' );
	}
	/**
	 * Setup plugin constants.
	 *
	 * @access private
	 * @since 1.0.0
	 * @return void
	 */
	private function setup_constants() {
		// Plugin version.
		if ( ! defined( 'SLIM_POPUPS_VERSION' ) ) {
			define( 'SLIM_POPUPS_VERSION', '1.0.2' );
		}
		// Plugin Folder Path.
		if ( ! defined( 'SLIM_POPUPS_PLUGIN_DIR' ) ) {
			define( 'SLIM_POPUPS_PLUGIN_DIR', plugin_dir_path( __FILE__ ) );
		}
		// Plugin Includes Path
		if ( ! defined( 'SLIM_POPUPS_INCLUDES_DIR' ) ) {
			define( 'SLIM_POPUPS_INCLUDES_DIR', SLIM_POPUPS_PLUGIN_DIR . 'includes/' );
		}
		// Plugin Folder URL.
		if ( ! defined( 'SLIM_POPUPS_PLUGIN_URL' ) ) {
			define( 'SLIM_POPUPS_PLUGIN_URL', plugin_dir_url( __FILE__ ) );
		}
		// Plugin Root File.
		if ( ! defined( 'SLIM_POPUPS_PLUGIN_FILE' ) ) {
			define( 'SLIM_POPUPS_PLUGIN_FILE', __FILE__ );
		}
		// Plugin Base Name
		if ( ! defined( 'SLIM_POPUPS_BASENAME' ) ) {
			define( 'SLIM_POPUPS_BASENAME', dirname( plugin_basename( __FILE__ ) ) );
		}
	}
	/**
	 * Include required files.
	 *
	 * @access private
	 * @since 1.0.0
	 * @return void
	 */
	private function includes() {
		// Vendor
		require_once SLIM_POPUPS_INCLUDES_DIR . 'lib/class-gamajo-template-loader.php';
		// Classes
		require_once SLIM_POPUPS_INCLUDES_DIR . 'class-template-loader.php';
		// Functions
		// require_once SLIM_POPUPS_INCLUDES_DIR . 'functions.php';
	}

	function setup() {
		register_activation_hook( __FILE__,   array( $this, 'activate' ) );
		register_deactivation_hook( __FILE__, array( $this, 'deactivate' ) );
		// Register styles and scripts
		add_action( 'wp_enqueue_scripts', array( $this, 'register_stylesheets' ) );
		add_action( 'wp_enqueue_scripts', array( $this, 'register_scripts' ) );
	}

	function activate() {
	}

	function deactivate() {
	}

	/**
	 * Register stylesheets for later use
	 *
	 * Use via wp_enqueue_style('wampum'); in a template
	 *
	 * @since  1.0.0
	 *
	 * @return null
	 */
	function register_stylesheets() {
	    wp_register_style( 'slim-popups', SLIM_POPUPS_PLUGIN_URL . 'css/slim-popups.css', array(), SLIM_POPUPS_VERSION );
	}

	/**
	 * Register scripts for later use
	 *
	 * Use via wp_enqueue_script('slim-popups'); in a template
	 *
	 * @since  1.0.0
	 *
	 * @return null
	 */
	function register_scripts() {
		wp_register_script( 'ouibounce', 	SLIM_POPUPS_PLUGIN_URL . 'js/ouibounce.min.js', array(), 			'0.0.12', 			  true );
		wp_register_script( 'slim-popups', 	SLIM_POPUPS_PLUGIN_URL . 'js/slim-popups.js', 	array('ouibounce'),  SLIM_POPUPS_VERSION, true );
	}

	function slim_popup( $filename, $options = array(), $args = array() ) {

		// Bail if no template (last parameter is whether or not to $load)
		$template = Slim_Popups()->templates->get_template_part( $filename, null, false );
		if ( ! $template ) {
			return;
		}

		// Popup options
		$defaults = array(
			'css'  			=> true, 	// whether or not to load the stylesheet
			'style'			=> 'modal', // 'modal' or 'slideup'
			'time'			=> '4000',  // time in milliseconds
			'type' 			=> 'exit',  // 'exit' or 'timed'
		);
		$options = wp_parse_args( $options, $defaults );

		wp_enqueue_script('ouibounce');
		wp_enqueue_script('slim-popups');

		$this->localize_script( $options, $args );

		if ( $options['css'] ) {
			wp_enqueue_style('slim-popups');
		}

		echo '<div id="slim-popups" style="display:none;">';
			echo '<div class="slim-popups-overlay">';
				echo '<div class="slim-popup">';
					echo '<div class="slim-popup-close"><button>×<span class="screen-reader-text">Close Popup</span></button></div>';
					echo '<div class="slim-popup-content">';
					    Slim_Popups()->templates->get_template_part( $filename, null, true );
					echo '</div>';
				echo '</div>';
			echo '</div>';
		echo '</div>';
	}

	function localize_script( $options, $args ) {
		$array = array(
			'slimpopups' => $options,
			'ouibounce'  => $this->ouibounce_args( $args ),
		);
		wp_localize_script( 'slim-popups', 'slim_popups_vars', $array );
	}

	function ouibounce_args( $args ) {
		// Script defaults
		$defaults = array(
			'aggressive'	=> false,   // true
			'callback'		=> false,   // function() { console.log('slim popups fired!'); }
			'cookieexpire'	=> false,   // 7
			'cookiedomain'	=> false,   // .example.com
			'cookiename'	=> false,   // 'custom_cookie_name'
			'delay'			=> false,   // 100
			'sensitivity'	=> false,   // 40
			'sitewide'		=> true,    // true (don't be annoying)
			'timer'			=> false,   // 10
		);
		$args  = wp_parse_args( $args, $defaults );
		$array = array();
		foreach ( $args as $key => $value ) {
			if ( $value == false ) {
				continue;
			}
			$array[$key] = $value;
		}
		return $array;
	}

}
endif; // End if class_exists check.

/**
 * The main function for that returns Slim_Popups_Setup
 *
 * The main function responsible for returning the one true Slim_Popups_Setup
 * Instance to functions everywhere.
 *
 * Use this function like you would a global variable, except without needing
 * to declare the global.
 *
 * Example: <?php $slim_popups = Slim_Popups_Setup_Init(); ?>
 *
 * @since 1.0.0
 *
 * @return object|Slim_Popups_Setup The one true Slim_Popups_Setup Instance.
 */
function Slim_Popups() {
	return Slim_Popups_Setup::instance();
}
// Get Slim_Popups_Setup Running.
Slim_Popups();


wampum-popups.php
<?php
/**
 * @package   Wampum_Popups_Setup
 * @author    BizBudding, INC <mike@bizbudding.com>
 * @license   GPL-2.0+
 * @link      http://bizbudding.com.com
 * @copyright 2016 BizBudding, INC
 *
 * @wordpress-plugin
 * Plugin Name:        Wampum - Popups
 * Description:        A lightweight but flexible WordPress popups plugin
 * Plugin URI:         https://github.com/bizbudding/wampum-popups
 * Author:             Mike Hemberger
 * Author URI:         https://bizbudding.com
 * Text Domain:        wampum-popups
 * License:            GPL-2.0+
 * License URI:        http://www.gnu.org/licenses/gpl-2.0.txt
 *
 * Version:            2.1.0
 *
 * GitHub Plugin URI:  https://github.com/bizbudding/wampum-popups
 * GitHub Branch:      master
 */

// Exit if accessed directly.
if ( ! defined( 'ABSPATH' ) ) exit;

/**
 * Main functions to create a popup
 *
 * @since  1.0.0
 *
 * @param  string  $content  The content to output in the popup
 * @param  array   $args     Plugin options (type, style, time)
 */
function wampum_popup( $content, $args = array() ) {
	echo Wampum_Popups()->get_wampum_popup( $content, $args );
}

function get_wampum_popup( $content, $args = array() ) {
	return Wampum_Popups()->get_wampum_popup( $content, $args );
}

/**
 * Main functions to create a direct link to launch a popup
 * Type can only be 'link' or 'button'
 *
 * @since  2.0.0
 *
 * @param  string  $content  The content to output in the popup
 * @param  array   $args     Plugin options (type & text only)
 */
function wampum_popup_link( $content, $args = array() ) {
	// Params are backwards cause shortcodes have optional $content second
	echo Wampum_Popups()->wampum_popup_shortcode( $args, $content );
}

function get_wampum_popup_link( $content, $args = array() ) {
	// Params are backwards cause shortcodes have optional $content second
	return Wampum_Popups()->wampum_popup_shortcode( $args, $content );
}

if ( ! class_exists( 'Wampum_Popups_Setup' ) ) :
/**
 * Main Wampum_Popups_Setup Class.
 *
 * @since 1.0.0
 */
final class Wampum_Popups_Setup {
	/**
	 * Singleton
	 * @var   Wampum_Popups_Setup The one true Wampum_Popups_Setup
	 * @since 1.0.0
	 */
	private static $instance;

	// Set popup counter
	private $popup_counter = 0;

	// Whether to use ouibounce or not
	private $ouibounce = false;

	// Set the wp_localize_script args variable
	private $localize_args = array();

	/**
	 * Main Wampum_Popups_Setup Instance.
	 *
	 * Insures that only one instance of Wampum_Popups_Setup exists in memory at any one
	 * time. Also prevents needing to define globals all over the place.
	 *
	 * @since   1.0.0
	 * @static  var array $instance
	 * @return  object | Wampum_Popups_Setup The one true Wampum_Popups_Setup
	 */
	public static function instance() {
		if ( ! isset( self::$instance ) ) {
			// Setup the setup
			self::$instance = new Wampum_Popups_Setup;
			// Methods
			self::$instance->setup_constants();
			self::$instance->setup();
		}
		return self::$instance;
	}
	/**
	 * Throw error on object clone.
	 *
	 * The whole idea of the singleton design pattern is that there is a single
	 * object therefore, we don't want the object to be cloned.
	 *
	 * @since   1.0.0
	 * @access  protected
	 * @return  void
	 */
	public function __clone() {
		// Cloning instances of the class is forbidden.
		_doing_it_wrong( __FUNCTION__, __( 'Cheatin&#8217; huh?', 'wampum' ), '1.0' );
	}
	/**
	 * Disable unserializing of the class.
	 *
	 * @since   1.0.0
	 * @access  protected
	 * @return  void
	 */
	public function __wakeup() {
		// Unserializing instances of the class is forbidden.
		_doing_it_wrong( __FUNCTION__, __( 'Cheatin&#8217; huh?', 'wampum' ), '1.0' );
	}
	/**
	 * Setup plugin constants.
	 *
	 * @access private
	 * @since 1.0.0
	 * @return void
	 */
	private function setup_constants() {
		// Plugin version.
		if ( ! defined( 'WAMPUM_POPUPS_VERSION' ) ) {
			define( 'WAMPUM_POPUPS_VERSION', '2.1.0' );
		}
		// Plugin Folder Path.
		if ( ! defined( 'WAMPUM_POPUPS_PLUGIN_DIR' ) ) {
			define( 'WAMPUM_POPUPS_PLUGIN_DIR', plugin_dir_path( __FILE__ ) );
		}
		// Plugin Includes Path
		if ( ! defined( 'WAMPUM_POPUPS_INCLUDES_DIR' ) ) {
			define( 'WAMPUM_POPUPS_INCLUDES_DIR', WAMPUM_POPUPS_PLUGIN_DIR . 'includes/' );
		}
		// Plugin Folder URL.
		if ( ! defined( 'WAMPUM_POPUPS_PLUGIN_URL' ) ) {
			define( 'WAMPUM_POPUPS_PLUGIN_URL', plugin_dir_url( __FILE__ ) );
		}
		// Plugin Root File.
		if ( ! defined( 'WAMPUM_POPUPS_PLUGIN_FILE' ) ) {
			define( 'WAMPUM_POPUPS_PLUGIN_FILE', __FILE__ );
		}
		// Plugin Base Name
		if ( ! defined( 'WAMPUM_POPUPS_BASENAME' ) ) {
			define( 'WAMPUM_POPUPS_BASENAME', dirname( plugin_basename( __FILE__ ) ) );
		}
	}

	/**
	 * Plugin hooks, filters, and shortcode
	 *
	 * @since  1.0.0
	 *
	 * @return void
	 */
	function setup() {

		// Register styles and scripts
		add_action( 'wp_enqueue_scripts', array( $this, 'stylesheets' ) );
		add_action( 'wp_enqueue_scripts', array( $this, 'scripts' ) );

		// Force large image url in all gallery links
		add_filter( 'wp_get_attachment_link', array( $this, 'gallery_image_url' ), 10, 4 );

		// Register our shortcode
		add_shortcode( 'wampum_popup', array( $this, 'wampum_popup_shortcode' ) );

		// Add new hook
		add_action( 'wampum_popups', array( $this, 'gallery_popup' ) );

		// Add our custom popup hook
		add_action( 'wp_footer', array( $this, 'popups_hook' ) );

		/**
		 * Setup the updater.
		 *
		 * @uses    https://github.com/YahnisElsts/plugin-update-checker/
		 *
		 * @return  void
		 */
		if ( ! class_exists( 'Puc_v4_Factory' ) ) {
			require_once WAMPUM_POPUPS_INCLUDES_DIR . 'vendor/plugin-update-checker/plugin-update-checker.php';
		}
		$updater = Puc_v4_Factory::buildUpdateChecker( 'https://github.com/bizbudding/wampum-popups/', __FILE__, 'wampum-popups' );
	}

	/**
	 * Register stylesheets for later use
	 *
	 * Use via wp_enqueue_style('wampum'); in a template
	 *
	 * @since  1.0.0
	 *
	 * @return null
	 */
	function stylesheets() {
	    wp_register_style( 'wampum-popups', WAMPUM_POPUPS_PLUGIN_URL . 'css/wampum-popups.min.css', array(), WAMPUM_POPUPS_VERSION );
	}

	/**
	 * Register scripts for later use
	 *
	 * Use via wp_enqueue_script('wampum-popups'); in a template
	 *
	 * @since  1.0.0
	 *
	 * @return null
	 */
	function scripts() {
		wp_register_script( 'ouibounce', WAMPUM_POPUPS_PLUGIN_URL . 'js/ouibounce.min.js', array(), '0.0.12', true );
		wp_register_script( 'wampum-popups', WAMPUM_POPUPS_PLUGIN_URL . 'js/wampum-popups.min.js', 	array('jquery'), WAMPUM_POPUPS_VERSION, true );
	}

	/**
	 * Force large image url in all gallery links
	 *
	 * @since   2.0.0
	 *
	 * @return  string  Image URL
	 */
	function gallery_image_url( $html, $post_id, $size, $permalink ) {
		$size  = apply_filters( 'wampum_popups_gallery_image_size', 'large' );
		$image = wp_get_attachment_image_src( $post_id, $size );
		return preg_replace( '/href=\'(.*?)\'/', 'href=\'' . $image[0] . '\'', $html );
	}

	/**
	 * See get_wampum_popup()
	 * This is only here to handle the shortcode
	 *
	 * @since  2.0.0
	 */
	function wampum_popup_shortcode( $atts, $content = null ) {

		// Get popup content, while parsing shortcodes
		$popup_content = do_shortcode(trim($content));

		// Bail if no content
		if ( empty( $popup_content ) ) {
			return;
		}

		// Get popup as a variable so it updates popup_counter
		$popup = $this->get_wampum_popup( $popup_content, $atts );

		// Bail if no popup
		if ( ! $popup ) {
			return;
		}

		/**
		 * Echo popup to footer so it's not inline and weird
		 * If inline, it was also too aggressively adopting .entry-content related styles
		 */
		add_action( 'wampum_popups', function() use ( $popup ) {
			echo $popup;
		});

		// If popup is a link or button type, return them linked text
		$output = '';
		$click  = array('link','button');
		if ( in_array( $atts['type'], $click ) ) {
			if ( isset($atts['text']) && ! empty($atts['text']) ) {
				if ( 'link' == $atts['type'] ) {
					$classes = 'wampum-popup-link';
				} elseif ( 'button' == $atts['type'] ) {
					$classes = 'wampum-popup-link button';
				}
				$output = sprintf( '<a class="%s" href="#" data-popup="%s">%s</a>',
					$classes,
					$this->popup_counter,
					sanitize_text_field($atts['text'])
				);
			}
		}
		return $output;
	}

	/**
	 * Get a wampum popup
	 * Enqueues scripts
	 * Returns the popup so we can use it in a shortcode
	 *
	 * @since  1.0.2
	 *
	 * @param  string  $content  HTML to be used in the popup
	 * @param  array   $args     All the popup args
	 *
	 * @return string  popup HTML
	 */
	function get_wampum_popup( $content = null, $args = array() ) {

		// Bail if an exit or timed popup and has no content
		if ( in_array( $args['type'], array('exit','timed') ) && ! trim($content) ) {
			return;
		}

		// Popup args
		$defaults = array(
			'type'          => null,    // 'exit' or 'timed' or 'link' or 'button' or 'gallery (internal only)' (REQUIRED)
			'close_button'  => true,    // whether or not to show the close button
			'close_outside' => true,    // whether or not to allow close by clicking outside the modal
			'logged_in'     => false,   // whether or not to show only to logged in users
			'logged_out'    => false,   // whether or not to show only to logged out users
			'style'         => 'modal', // 'modal' or 'slideup'
			'time'          => '4000',  // time in milliseconds
			'width'         => '400px', // max-width of popup
			'aggressive'    => false,   // ouibounce - true
			'cookieExpire'  => false,   // ouibounce - 7
			'cookieDomain'  => false,   // ouibounce - .example.com
			'cookieName'    => 'wampumPopupViewed',   // ouibounce - 'custom_cookie_name'
			'delay'         => false,   // ouibounce - 100
			'sensitivity'   => false,   // ouibounce - 40
			'sitewide'      => true,    // ouibounce - true (don't be annoying)
			'timer'         => false,   // ouibounce - 10
		);
		$args = shortcode_atts( $defaults, $args, 'wampum_popup' );

		// Bail if we don't have a type, since it's required!
		$types = array('link','button','exit','gallery','timed');
		if ( ! in_array( $args['type'], $types ) ) {
			return;
		}

		// Bail if logged_in is true and user is not logged in
		if ( $args['logged_in'] && ! is_user_logged_in() ) {
			return;
		}

		// Bail if logged_out is true and user is not logged in
		if ( $args['logged_out'] && is_user_logged_in() ) {
			return;
		}

		// Array of types that use ouibounce
		$ouibounce = array('exit','timed');

		// If an ouibounce popup
		if ( in_array( $args['type'], $ouibounce ) ) {

			// Bail if popup is not aggressive and cookie has already been viewed
			$aggressive = filter_var( $args['aggressive'], FILTER_VALIDATE_BOOLEAN );
			$viewed     = isset($_COOKIE[$args['cookieName']]) && filter_var( $_COOKIE[$args['cookieName']], FILTER_VALIDATE_BOOLEAN ) == true ? true : false;
			if ( ! $aggressive && $viewed ) {
				return;
			}

			// If popup uses ouibounce, set as true
			$this->ouibounce = true;

		}

		// Increment the counter so JS can fire the correct popup if multiple on the same page!
		$this->popup_counter++;

		/**
		 * Add these args to a big localization array that saves each popup in its own index
		 * First get the array of this specific popup's args
		 * Then add this as an index to the main localize object
		 */
		$localize_args = $this->get_localize_script_args( $args );
		$this->localize_args[$this->popup_counter] = $localize_args;

		// Send it!
		return $this->get_wampum_popup_html( $content, $args, $this->popup_counter );

	}

	function get_wampum_popup_html( $content, $args, $index = null ) {
		// Maybe add close outside class
		$close_outside = filter_var( $args['close_outside'], FILTER_VALIDATE_BOOLEAN ) ? ' close-outside' : '';
		// The markup
		$output  = '';
		$output .= '<div id="wampum-popup-' . $index . '" class="wampum-popup" style="display:none;" data-popup="' . $index . '">';
			$output .= '<div class="wampum-popup-overlay' . $close_outside . '">';
				$output .= '<div class="wampum-popup-inner" style="max-width:' . $args['width'] . ';">';
					// If close button, add it
					if ( filter_var( $args['close_button'], FILTER_VALIDATE_BOOLEAN ) ) {
						$output .= '<div class="wampum-popup-button wampum-popup-close"><span class="screen-reader-text">Close Popup</span></div>';
					}
					$output .= '<div class="wampum-popup-content">';
					    $output .= $content;
					$output .= '</div>';
				$output .= '</div>';
			$output .= '</div>';
		$output .= '</div>';
		return $output;
	}

	/**
	 * Get the properly organized array of args to localize
	 *
	 * @since  2.0.0
	 *
	 * @param  array  $args  Accepted args to localize
	 *
	 * @return array
	 */
	function get_localize_script_args( $args ) {
		$popup_args = array(
			'close_button'  => $args['close_button'],
			'close_outside' => $args['close_outside'],
			'style'         => $args['style'],
			'time'          => $args['time'],
			'type'          => $args['type'],
			'width'         => $args['width'],
		);
		$ouibounce_args = array(
			'aggressive'   => $args['aggressive'],
			'cookieExpire' => $args['cookieExpire'],
			'cookieDomain' => $args['cookieDomain'],
			'cookieName'   => $args['cookieName'],
			'delay'        => $args['delay'],
			'sensitivity'  => $args['sensitivity'],
			'sitewide'     => $args['sitewide'],
			'timer'        => $args['timer'],
		);
		return array(
			'wampumpopups' => $popup_args,
			'ouibounce'    => $this->ouibounce_args( $ouibounce_args ),
		);
	}

	/**
	 * Strip out the args we don't want to send to the ouibounce() function
	 *
	 * @since   1.0.2
	 *
	 * @param   array  $ouibounce_args  Array of args to check agains
	 *
	 * @return  array  The args to keep
	 */
	function ouibounce_args( $ouibounce_args ) {
		$args = array();
		// Script defaults
		$defaults = array(
			'aggressive'   => false,   // true
			'cookieExpire' => false,   // 7
			'cookieDomain' => false,   // .example.com
			'cookieName'   => false,   // 'custom_cookie_name'
			'delay'        => false,   // 100
			'sensitivity'  => false,   // 40
			'sitewide'     => true,    // true (don't be annoying)
			'timer'        => false,   // 10
		);
		$ouibounce_args	= wp_parse_args( $ouibounce_args, $defaults );
		foreach ( $ouibounce_args as $key => $value ) {
			if ( $value == false ) {
				continue;
			}
			$args[$key] = $value;
		}
		return $args;
	}

	function gallery_popup() {
		// If not a singular view and no gallery
		if ( ! ( is_singular() && get_post_gallery( get_the_ID(), false ) ) ) {
			return;
		}
		// Output popup with empty span so it forces display
		$args = array(
			'aggressive'    => true,
			'close_outside' => false,
			'style'         => 'modal',
			'type'          => 'gallery',
			'width'         => 'auto',
		);
		wampum_popup( '', $args );
	}

	/**
	 * Add a new hook so devs can safely add a new popup without things breaking if this plugin gets deactivated
	 *
	 * @since   1.1.0
	 *
	 * @return  null
	 */
	function popups_hook() {

		do_action( 'wampum_popups' );

		// If we have at least 1 popup
		if ( ( $this->popup_counter > 0 ) ) {
			$css = apply_filters( 'wampum_popups_load_css', '__return_true' );
			if ( $css ) {
				// Styles baby
				wp_enqueue_style( 'wampum-popups' );
			}
			// Scripts baby
			if ( $this->ouibounce ) {
				wp_enqueue_script( 'ouibounce' );
			}
			wp_enqueue_script( 'wampum-popups' );
			wp_localize_script( 'wampum-popups', 'wampum_popups_vars', $this->localize_args );
		}

	}

}
endif; // End if class_exists check.

/**
 * The main function for that returns Wampum_Popups_Setup
 *
 * The main function responsible for returning the one true Wampum_Popups_Setup
 * Instance to functions everywhere.
 *
 * Use this function like you would a global variable, except without needing
 * to declare the global.
 *
 * Example: <?php $wampum_popups = Wampum_Popups_Setup_Init(); ?>
 *
 * @since 1.0.0
 *
 * @return object|Wampum_Popups_Setup The one true Wampum_Popups_Setup Instance.
 */
function Wampum_Popups() {
	return Wampum_Popups_Setup::instance();
}

// Get Wampum_Popups_Setup Running.
Wampum_Popups();
